### PR TITLE
Fix for encodings listed in the Parquet column chunk metadata

### PR DESCRIPTION
**Pull request by etseidl**
## Description
With the addition of V2 page headers, the encodings used have also changed. This PR correctly determines the encodings used in each column chunk and writes that information to the column chunk metadata.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/writer_impl.cu [hunk 1]
```diff
@@ -1703,17 +1731,6 @@ auto convert_table_to_parquet_data(table_input_metadata& table_meta,
   row_group_fragments.host_to_device_async(stream);
   [[maybe_unused]] auto dict_info_owner = build_chunk_dictionaries(
     chunks, col_desc, row_group_fragments, compression, dict_policy, max_dictionary_size, stream);
-  for (size_t p = 0; p < partitions.size(); p++) {
-    for (int rg = 0; rg < num_rg_in_part[p]; rg++) {
-      size_t global_rg = global_rowgroup_base[p] + rg;
-      for (int col = 0; col < num_columns; col++) {
-        if (chunks.host_view()[rg][col].use_dictionary) {
-          agg_meta->file(p).row_groups[global_rg].columns[col].meta_data.encodings.push_back(
```
Comment (vuule) [inline cpp/src/io/parquet/writer_impl.cu:1711]: Is `column_chunk_meta.encodings` somehow the same as `agg_meta->file(p).row_groups[global_rg].columns[col].meta_data.encodings`?

File: cpp/tests/io/parquet_test.cpp [hunk 2]
```diff
@@ -6599,4 +6599,77 @@ TEST_F(ParquetWriterTest, TimestampMicrosINT96NoOverflow)
...
+    for (size_t i = 0; i < vec.size(); i++) {
+      if (vec[i] == enc) return true;
+    }
+    return false;
+  };
+
+  auto const source = cudf::io::datasource::create(filepath);
+  cudf::io::parquet::FileMetaData fmd;
+
+  read_footer(source, &fmd);
```
Comment (vuule) [inline cpp/tests/io/parquet_test.cpp:6642]: You KNOW the PR is in the weeds when the test calls `read_footer`!

**Review by vuule — COMMENTED** [review]
Big improvement on the way the encodings are handled 👍 
What does this fix impact?

File: cpp/src/io/parquet/page_enc.cu [hunk 3]
```diff
@@ -229,6 +229,16 @@ Encoding __device__ determine_encoding(PageType page_type,
   }
 }
 
+// operator to use with warp_reduce. stolen from cub::Sum
+struct Or {
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/page_enc.cu:233]: ok

File: cpp/tests/io/parquet_test.cpp [hunk 4]
```diff
@@ -6599,4 +6599,77 @@ TEST_F(ParquetWriterTest, TimestampMicrosINT96NoOverflow)
...
+      .write_v2_headers(is_v2);
+  cudf::io::write_parquet(out_opts);
+
+  // make sure the expected encodings are present
+  auto contains = [](auto vec, auto enc) {
+    for (size_t i = 0; i < vec.size(); i++) {
+      if (vec[i] == enc) return true;
+    }
+    return false;
+  };
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/tests/io/parquet_test.cpp:6637]: !#@$? C++.  Was looking for `find` as a member of `vector` 😅

File: cpp/src/io/parquet/parquet_gpu.hpp [hunk 5]
```diff
@@ -365,6 +365,28 @@ constexpr size_t kDictScratchSize    = (1 << kDictHashBits) * sizeof(uint32_t);
 struct EncPage;
 struct slot_type;
 
+// convert Encoding to a mask value
+constexpr uint32_t encoding_to_mask(Encoding encoding)
+{
+  return 1 << static_cast<uint32_t>(encoding);
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/parquet_gpu.hpp:371]: Given it's passed an enum, this should be safe without the check.  Could add a max value to the Encoding enum, but it'll take a long time to get 22 more encodings in the spec 🤣

**Comment by etseidl** [comment]
> What does this fix impact?

Just correctness.  As I dump files it's annoying to see encodings listed that aren't used.  And this will help when the delta encodings come online.

**Review by vuule — APPROVED** [review]
Looks great, love the version without `EncodingMask`!

**Review by ttnghia — APPROVED** [review]

**Pull request merged**